### ISSUE TITLE
SamplerAutoConfiguration conflicts with TraceAutoConfiguration

### ISSUE THREAD

**Issue by snicoll** [spring-cloud/spring-cloud-sleuth#1618]
**Describe the bug**
Spring Cloud Sleuth provides two auto-configurations that are competing with each other. `TraceAutoConfiguration` and `SamplerAutoConfiguration` both provides a `Sampler` but there is no order of precedence or excluding conditions whatsoever.  They both applies and the order of execution determines who wins.

This makes it for an odd situation as the former provides a default bean that doesn't do any tracing at all and the latter provides a default sampler bean that is configured according to configuration. `TraceAutoConfiguration` wins for the sole reason that its package is lexicographically before `SamplerAutoConfiguration`. 

Looking at the codebase, it looks like a common practice is to provide a `Sampler` bean to enable this feature. I'd argue that `SamplerAutoConfiguration` should not be an auto-configuration then. Or rather, if it is, it should override the default behaviour of `TraceAutoConfiguration` or offer a sensible condition to disable itself.

`SamplerAutoConfiguration` has some logic to create a sensible sampler based on configuration properties defined in `spring.sleuth.sampler.*`. Unfortunately, a user can't replicate that behaviour as part of the code is package private.

I don't know what the best outcome of this one is based on constraints but if the plan is to disable the `Sampler` bean by default, then I think we should expose a public API to let you create a default sampler (basically what `SamplerAutoConfiguration` does at the moment). Users could then inject that factory and create a sampler or ignore it altogether and just create the bean themselves if they want to.

**Comment by adriancole** [comment]
I don't expect anyone but me to answer this as marcin is out and everyone  else are swamped also :)

I think it should never have been an auto-configuration. there's some other glitches around this also. we should have just imported it, I think.

**Comment by adriancole** [comment]
fixed in 2.2.x via https://github.com/spring-cloud/spring-cloud-sleuth/commit/16fb8e3ac6cac79d549251b9d87841a3ff180974 and will do that in the other branches

**Comment by adriancole** [comment]
FWIW, this seems to have been a relic of 2.0 where zipkin held the sampler configuration, but was in a different jar file. Basically, it was only partially corrected in 2.1.x (until now)

Thanks for looking deeply. The next version of sleuth will correct the odd dep, though I would advise that unless you can depend on a not-yet-released version of sleuth, import `SamplerAutoConfiguration` directly, as not doing this makes things default in a strange way that's very hard to understand.

**Comment by snicoll** [comment]
Thanks for the quick reply. Unfortunately, a regular configuration should not be using `@ConditionalOnMissingBean` (see [the javadoc](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/autoconfigure/condition/ConditionalOnMissingBean.html)). 

With this change the no_sampler `sleuthTraceSampler` will never be created  as the two inner classes are mutually exclusive and will create one or the other.

Can we please reopen this one to consider other options? Thanks!

**Comment by adriancole** [comment]
@snicoll if what you are saying is true, our tests are horrible as they all pass

**Comment by adriancole** [comment]
actually this is a non-problem. I'll remove that default to no sampler agree it is distracting. this whole code base is a disaster

**Comment by adriancole** [comment]
I'm doing a big doc cleanup now.

There was a large amount of confusion in the docs and code around this.

There's no relevance to sampling in logs as they are always put with IDs.  The only relevance to sampling when logging is that you are *not* only making a sampling decision if a backend is configured in some way.

The focus should be on a tracing backend. Before, the config would read sampling properties only if zipkin was in the classpath. The change I made fixed that and this made the default that was in the tracer unread, as you noticed, and there was no test to catch it.

This is all about the odd-case where sleuth is used in a log-only environment, and I suppose there is fear that somewhere down the line there will be a zipkin who cannot control sampling because it was already set by the log-only ones.

I'll try my best to unwind this.

**Comment by adriancole** [comment]
here's the main doc I will add, which is complicated because of the log-only, but since we have supported it in the past we unfortunately need to explain it.

---
Sampling only applies to tracing backends, such as Zipkin. Trace IDs appear in logs regardless of sample rate. Sampling is a way to prevent overloading the system, by consistently tracing some, but not all requests.

The default rate of 10 traces per second is controlled by the `spring.sleuth.sampler.rate`
property and applies when we know Sleuth is used for reasons besides logging.

Specifically, we look for presence of `FinishedSpanHandler` or `Reporter<Span>` beans as a signal
that tracing is in use. In the odd case you need to override this heuristic to never sample,
set `spring.sleuth.sampler.rate=0`.

The rate should reflect reality of your service, limited to your tracing service capacity. For
example, `spring.sleuth.sampler.rate` is an integer property, so the max possible is 2147483647. However, that's a rate per second. It is unlikely that a single process could service over 2 million requests. The largest known tracing sites can process 1M spans per second for the entire infrastructure. A trace usually has multiple spans, and can sometimes have a thousand. Considering the above, single-process rates above a thousand are more likely to flood most systems vs provide value.

In short, use rates above 100 with extreme caution.

**Comment by adriancole** [comment]
https://github.com/spring-cloud/spring-cloud-sleuth/pull/1619 puts back the weird thing, though we should remove it in 3.0 as it is completely unintuitive

**Comment by snicoll** [comment]
Thanks for the follow-up Adrian. Given our constraints of having to use 2.2, let's use the workaround indeed, we can revisit that soon enough.

**Comment by marcingrzejszczak** [comment]
So what's the state of this? For 2.2 we leave things as they are and for 3.x we want to have the default sampler as the sampler properties based one (regardless of whether we have logging only / Zipkin / Wavefront)? That would make sense.

**Comment by adriancole** [comment]
current state is sampler is the same 2.2.x as 3.0.x

People shouldn't import SamplerAutoConfiguration anymore. Those already setting a sampler (as they did before) can continue to do that.

Ex the following non-default things don't import sampler anymore

https://github.com/wavefrontHQ/wavefront-spring-boot/pull/55/files#diff-894100fa6d9a55386f4a6d9463676b44L32
https://github.com/spring-cloud/spring-cloud-gcp/pull/2290/files#diff-fc1b9a6100070e7eb2d92338e77686e5L84
